pkg/health/checker/checker.go

```diff
@@ -83,41 +83,95 @@ func (c consulHealthChecker) WatchNodeService(
 	}
 }
 
+// publishLatestHealth is not thread safe - do not start more than one of these per resultCH
+func publishLatestHealth(inCh <-chan api.KVPairs, quitCh <-chan struct{}, resultCh chan []*health.Result) chan error {
+	errCh := make(chan error)
+
+	go func() {
+		var listed api.KVPairs
+		var ok bool
+		var err error
+
+		for {
+			// We don't have a value, fetch a fresh one or skip
+			select {
+			case listed, ok = <-inCh:
+				if !ok {
+					// channel closed
+					return
+				}
+			case <-quitCh:
+				return
+			}
+
+			results := make([]*health.Result, 0, len(listed)) // allocate a new return slice for each watch
+			results, err = kvpsToResult(listed)
+			if err != nil {
+				select {
+				case errCh <- err:
+					// The most recent update is an error.
+					// We go back to the start in this case
+					continue
+				case <-quitCh:
+					return
+				}
+			}
+
+			// here, we prepare to write the value.
+			// First we drain the resultChan of any stale health results
+			select {
+			case _, ok = <-resultCh:
+				if !ok {
+					return
+				}
+			default:
+			}
+			// Now we check the quit chan and try to write to our resultCh
+			select {
+			case <-quitCh:
+				return
+			case resultCh <- results:
+			}
+		}
+	}()
+
+	return errCh
+}
+
 // Watch the health tree and write the whole subtree on the chan passed by caller
+// the result channel argument _must be buffered_
+// Any errors are passed, best effort, over errCh
 func (c consulHealthChecker) WatchHealth(
-	resultCh chan<- []*health.Result,
+	resultCh chan []*health.Result,
 	errCh chan<- error,
 	quitCh <-chan struct{},
 ) {
-	defer close(resultCh)
-
-	res := make(chan api.KVPairs)
-	defer close(res)
-
-	quitWatch := make(<-chan struct{})
-
-	go consulutil.WatchPrefix("health/", c.kv, res, quitWatch, errCh)
+	// closed by watchPrefix when we close quitWatch
+	inCh := make(chan api.KVPairs)
+	watchErrCh := make(chan error)
+	go consulutil.WatchPrefix("health/", c.kv, inCh, quitCh, watchErrCh)
+	publishErrCh := publishLatestHealth(inCh, quitCh, resultCh)
 
-	var results api.KVPairs
-	for {
+	go func() {
```

Comment:
nit: i'd prefer for this function to block since it's easy for a caller to call it in a goroutine if they so desire. @anthonybishopric once showed me a talk asserting that libraries should always be blocking